src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -140,81 +141,132 @@ initialScannerState = ScannerState
-removeIgnored fp = withIgnoreMode . cataNodeWithParentNodeInfo remove
+-- and remove nodes that should be ignored.
+processAnnotations :: FilePath -> C.Node -> Writer [ScanError] C.Node
+processAnnotations fp = withIgnoreMode . cataNodeWithParentNodeInfo process
   where
-    remove
+    process
       :: Maybe PosInfo
       -> NodeType
-      -> [ScannerM Node]
-      -> ScannerM Node
-    remove pos ty subs = do
-      let node = Node pos ty []
-      scan <- use ssIgnore >>= \e -> do
+      -> [ScannerM C.Node]
+      -> ScannerM C.Node
+    process pos ty subs = do
+      let node = C.Node pos ty []
+      use ssIgnore >>= \ign -> do
         -- When no `Ignore` state is set check next node for annotation,
         -- if found then set it as new `IgnoreMode` otherwise skip node.
         let mbAnnotation = getAnnotation node
         case mbAnnotation of
           Just ann -> handleAnnotation pos ty ann
-          Nothing -> case e of
-            Nothing                    -> Node pos ty <$> sequence subs
-            Just (Ignore mode modePos) ->
-                case (mode, ty) of
-                  -- We expect to find a paragraph immediately after the
-                  -- `ignore paragraph` annotanion. If the paragraph is not
-                  -- found we should report an error.
-                  (IMSParagraph, PARAGRAPH)    -> (ssIgnore .= Nothing) $> defNode
-                  (IMSParagraph, x)            -> do
-                    lift . tell . makeError modePos fp . ParagraphErr $ prettyType x
-                    ssIgnore .= Nothing
-                    Node pos ty <$> sequence subs
-
-                  (IMSLink _, LINK {})         -> do
-                    ssIgnore .= Nothing
-                    return defNode
-                  (IMSLink _, IMAGE {})        -> do
-                    ssIgnore .= Nothing
-                    return defNode
-                  (IMSLink ignoreLinkState, _) -> do
-                    when (ignoreLinkState == ExpectingLinkInSubnodes) $
-                      ssIgnore . _Just . ignoreMode .=  IMSLink ParentExpectsLink
-                    node' <- Node pos ty <$> sequence subs
-                    when (ignoreLinkState == ExpectingLinkInSubnodes) $ do
-                      currentIgnore <- use ssIgnore
-                      case currentIgnore of
-                        Just (Ignore {_ignoreMode = IMSLink ParentExpectsLink}) -> do
-                          lift $ tell $ makeError modePos fp LinkErr
-                          ssIgnore .= Nothing
-                        _ -> pass
-                    return node'
-
-      when (ty == PARAGRAPH) $ use ssIgnore >>= \case
+          Nothing -> do
+            case ty of
+              PARAGRAPH -> handleParagraph ign pos ty subs
+              LINK {}   -> handleLink      ign pos ty subs
+              IMAGE {}  -> handleLink      ign pos ty subs
+              _         -> handleOther     ign pos ty subs
```

Comment:
For me, it's easier to reason about code behavior when a node type is known, because our types for ignore modes are, IMHO, rather confusing and hard to understand. And, in contrast, node type is something very clear and easy to work with, so I preferred it.

Things only gets worse if there are some different types of annotation (the primary reason of the whole this refactor). We'll have to handle a lot of different cases for all types of annotations and nodes.